scRNA-seq-advanced/01-reading_filtering_CellRanger.Rmd

```diff
@@ -0,0 +1,484 @@
+
+
+### Filtering empty droplets
+
+Most of the barcodes in any given 10X experiment will not be seen at all, so our first step can be to filter this raw data to only the cells where there is at least one transcript that was counted with that barcode.
+
+To do this, we will use the `colSums()` function to quickly add up all the counts that correspond to each possible barcode, then filter our `raw_sce` down to just those columns.
+We will need to extract the `counts` matrix from our SCE object, which we can do using the `counts()` function, conveniently enough.
+
+```{r remove zeros}
+# sum columns from counts matrix
+barcode_counts <- colSums(counts(raw_sce))
+
+# filter SCE object to only rows with counts > 0
+raw_sce <- raw_sce[, which(barcode_counts > 0)]
+```
+
+Now we can look at how our SCE object has changed:
+
+```{r zero-removed SCE}
+raw_sce
+```
+
+Of course, barcodes with zero counts are not the only ones that will not correspond to droplets without cells in them. 
+Even if a cell does not have a cell in it, there will often be spurious reads from RNA sequences that were present in the extracellular solution, whether from the original sample or cells that were disrupted during preparation for single-cell analysis. 
+
+We can identify these in part by low transcript counts, but we can also be a bit more clever, and look at the transcript counts _from_ the lowest-count droplets to create an expected distribution of transcript counts in droplets that don't contain cells.
+Then we can test each cell to determine whether or not its transcript distribution deviates from that expectation.
+If it does, then we have pretty good evidence that there is a cell in there.
+
+This test was first proposed by [Lun _et al._ (2019)](https://doi.org/10.1186/s13059-019-1662-y) and implemented as `emptyDrops()` in the `DropletUtils` package. 
+This method was then adopted, with some modifications, as the default cell filtering method used by Cell Ranger.
+Here we will use the `emptyDropsCellRanger()` function to perform the filtering to most closely match the Cell Ranger implementation.
+
+
+```{r calculate droplet stats}
+# create a table of statistics using emptyDropsCellRanger
+droplet_df <- DropletUtils::emptyDropsCellRanger(counts(raw_sce))
+```
+
+If we look at this table, we will see that most of the values in it are `NA`, because the low-count cells are used to generate the background distribution, and so do not have individual statistics calculated.
+
+We can look at just the rows without `NA` values by selected the ones where the FDR (which we will use again soon), is not `NA`
+
+```{r look at droplet stats}
+# view rows where FDR is not `NA`
+droplet_df[!is.na(droplet_df$FDR), ]
+```
+You will notice that some cells with high counts also have NA values for many statistics, but in those cases  it is because they are automatically assumed to contain cells, so were not tested. 
```

Comment:
This is such a "weird" feature that I tried to make the text a bit more explicit about why is this happening. take it or leave it, or something in between!

```suggestion
You will notice that some cells with high counts also have `NA` values for many statistics.
In those cases, `NA` values are actually present _because_ of the high counts - `emptyDropsCellRanger()` automatically assumed cells were present, so they were also not tested. 
```